doc/driver.mld

```diff
@@ -0,0 +1,377 @@
+can only read and modify the code locally: the parts of the AST that are given
+to the transformation are only portions of the whole AST. In this regard, they
+are usually called {e context-free} transformations. While being not as
+general-purpose as plain AST transformations, they are more than often
+sufficient and have many nice properties such as a well-defined semantic for
+composition.
+The two most important of such context-free transformations are {e derivers} and
+{e extenders}.
+
+{2:def_derivers Derivers}
+
+A {e deriver} is a context-free transformation that, given a structure or
+signature item, will generate code {e to append after} this item. The given code is
+never modified. A deriver can be very useful to generate values depending on the
+structure of a user-defined type, for instance a converter of a value of a type
+to and from a JSON value. A deriver is triggered by adding an
+{{:https://v2.ocaml.org/manual/attributes.html}attribute} to a structure or
+signature item. For instance, the folowing code:
+
+{@ocaml[
+  type t = Int of int | Float of float [@@deriving show]
+
+  let x = ...
+]}
+
+would be rewritten to:
+
+{@ocaml[
+  type ty = Int of int | Float of float [@@deriving show]
+
+  let ty_of_yojson = ...
+  let ty_to_yojson = ...
+
+  let x = ...
+]}
+
+{2:def_extenders Extenders}
+
+An {e extender} is a context-free transformation that is triggered on
+{{:https://v2.ocaml.org/manual/extensionnodes.html}extension nodes}, and that
+will replace the extension node by some code generated from the payload of the
+extension node. This can be very useful to generate values of a DSL using a more
+user-friendly syntax, for instance to generate OCaml values from the json
+syntax.
+
+For instance, the folowing code:
+
+{@ocaml[
+  let json =
+    [%yojson
+      [ { name = "Anne"; grades = ["A"; "B-"; "B+"] }
+      ; { name = "Bernard"; grades = ["B+"; "A"; "B-"] }
+      ]
+    ]
+]}
+
+could be rewritten into:
+
+{@ocaml[
+  let json =
+    `List
+      [ `Assoc
+          [ ("name", `String "Anne")
+          ; ("grades", `List [`String "A"; `String "B-"; `String "B+"])
+          ]
+      ; `Assoc
+          [ ("name", `String "Bernard")
+          ; ("grades", `List [`String "B+"; `String "A"; `String "B-"])
+          ]
+      ]
+]}
+
+{2 Advantages}
+
+The advantages of using context-free transformations are multiple. First, they
+provide to the user a much clearer understanding of the parts of the AST that
+will be rewritten, than a fully general AST rewriting. Secondly, they provide
+a much better composition semantic, which does not depend on the order. Finally,
+context-free transformation are applied in a single phase factorising the work
+for all transformation, resulting in a much faster driver than when combining
+multiple whole AST transformations. More details on the execution of this phase
+are given in the {{!"context-free-phase"}dedicated section.}
```

Comment:
I add back that the advantage is for the PPX user (as opposed to the PPX writer)
```suggestion
There are multiple advantages of using context-free transformations. First, they
provide the PPX user a much clearer understanding of the AST parts that
will be rewritten, rather than a fully general AST rewriting. Secondly, they provide
a much better composition semantic, which does not depend on the order. Finally,
context-free transformations are applied in a single phase factorising the work
for all transformations, resulting in a much faster driver than when combining
multiple, whole AST transformations. More details on the execution of this phase
are given in its {{!"context-free-phase"}dedicated section.}
```